### PR TITLE
Allow users to take GCS-based backups

### PR DESCRIPTION
**Pull request by gerlowskija**
This commit adds first-pass support for exposing Solr's
GcsBackupRepository through our operator configuration.  This WIP
support has a number of caveats and downsides that'll need agreed
on or fixed prior to merging :

  - GCS backups eschew the "persistence" step that currently follows
    normal backups
  - GCS backups are only included in Solr 8.9+, but there's no check for
    this currently.
  - operator logic currently assumes that exactly 1 type of backup
    config will be provided on a given solrcloud object (i.e. GCS
    backups and 'local' PV backups are mutually exclusive for a
    solrcloud.
  - no automated tests have been added
  - no documentation of has been added, beyond the examples on issue
    #301

### CODE REVIEW

**Comment by gerlowskija** [comment]
I've updated this to use different CRD's (see the latest comment on the related issue #301).  This also adds support for defining multiple backup repositories in your solrcloud definition which is pretty cool.

Still needs tests, docs for the new CRD fields, and there's currently a bug where GCS backups succeed but this success is never persisted in the kube status.  Hope to fix those shortly.

File: controllers/solrbackup_controller.go [hunk 1]
```diff
@@ -253,7 +283,21 @@ func persistSolrCloudBackups(r *SolrBackupReconciler, backup *solrv1beta1.SolrBa
...
+	backupRepository, found := util.GetBackupRepositoryByName(solrCloud.Spec.StorageOptions.BackupRestoreOptions, backup.Spec.RepositoryName)
+	if ! found {
+		err = fmt.Errorf("Unable to find backup repository to use for backup [%s] (which specified the repository" +
+			"[%s]).  solrcloud must define a repository matching that name (or have only 1 repository defined).",
+			backup.Name, backup.Spec.RepositoryName)
+		return err
+	}
+
+	managedBackupRepository, ok := backupRepository.(util.ManagedBackupRepository)
+	if ! ok {
```
Comment (HoustonPutman) [inline controllers/solrbackup_controller.go:295]: I'd rather the persistence stuff go in this `if ok {`, otherwise it's a little hard to follow.

**Review by HoustonPutman — CHANGES_REQUESTED** [review]
Hey @gerlowskija , I think this is an awesome start and pretty far along. 

I've made some feedback, but I will definitely be helping more when the time permits. I'm not sure this will get into the v0.4.0 release, but hopefully we can have another release soon after.

File: api/v1beta1/solrbackup_types.go [hunk 2]
```diff
@@ -38,12 +38,16 @@ type SolrBackupSpec struct {
...
+	// auto-configured repository for legacy singleton volumes).
+	RepositoryName string `json:"repositoryName,omitempty"`
+
 	// The list of collections to backup. If empty, all collections in the cloud will be backed up.
 	// +optional
 	Collections []string `json:"collections,omitempty"`
 
 	// Persistence is the specification on how to persist the backup data.
-	Persistence PersistenceSource `json:"persistence"`
+	Persistence PersistenceSource `json:"persistence,omitempty"`
```
[the remark this answers is not in the document]
Comment (gerlowskija) [reply api/v1beta1/solrbackup_types.go:50]: (I hadn't seen this "Suggested Change" feature of Github's - very cool!)

I'm not against having a pointer here - I'll have to read the docs a bit on what you mean by "more optional" haha, but sounds good to me.

File: controllers/util/backup_util.go [hunk 3]
```diff
@@ -440,3 +483,13 @@ func RunExecForPod(podName string, namespace string, command []string, config re
 
 	return nil
 }
+
+// TODO Is the Mozilla license on hashicorp's go-version lib we're using here acceptable by Apache?  Is there a better option?
+func SupportsGcsBackups(versionStr string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (gerlowskija) [reply controllers/util/backup_util.go:488]: I'd need to double-check the codepath, but this doesn't use the image tag - it gets the data from Solr itself (I _think_ using `/solr/admin/info/system`)

**Comment by gerlowskija** [comment]
Thanks for the review Houston.  Just getting back from some time off now; gonna try to address your comments in the next day or two, as well as getting the docs and tests in order.

**Comment by gerlowskija** [comment]
Haven't touched this in a few weeks.  Both because of some competing priorities where I work, and because I've been scrambling to get things done before I disappear for an upcoming paternity leave.

But just wanted to say that this is still on my radar and I'm hoping to return to it as soon as I can.

**Comment by HoustonPutman** [comment]
> before I disappear for an upcoming paternity leave

Congrats!!! 🎉 

No worries, I've been busy on the S3 Repo stuff and the v0.4.0 release. I do plan on taking this forward over the next few weeks. Mind if I make some changes while you are working on other stuff?

File: api/v1beta1/solrcloud_types.go [hunk 4]
```diff
@@ -325,20 +329,158 @@ type SolrEphemeralDataStorageOptions struct {
...
+func (backupOptions *SolrBackupRestoreOptions) GetBackupPath(backupName string) string {
+	return backupOptions.GetSolrMountPath() + "/backups/" + backupName
+}
+
+func (backupOptions *SolrBackupRestoreOptions) GetVolumeName() string {
+	return BackupRestoreVolume
+}
+
 type SolrBackupRestoreOptions struct {
+	// TODO Do we need to support this (Volume) here for backcompat, or can it live exclusively in ManagedStorage?
```
[the remark this answers is not in the document]
Comment (gerlowskija) [reply api/v1beta1/solrcloud_types.go:361]: Just to double-check: when you say "auto-move" you're suggesting that early in its processing loop the operator would restructure and 'SolrBackupRestoreOptions' it sees so that any value held by 'Value' here is moved over to be a 'ManagedRepository' entry.  (And that way the rest of the operator can ignore the 'Volume' field).

File: controllers/util/backup_util.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerlowskija) [reply controllers/util/backup_util.go:488]: Just wanted to check in on your concerns here @HoustonPutman

[the remark this answers is not in the document]
Comment (gerlowskija) [reply controllers/util/backup_util.go:488]: Gah, nvm.  I just traced through where the Solr "version" comes from and you were correct that it's coming from the image tag and not anything more reliable.  Not sure why I thought it came from a Solr API iself.

This definitely needs to go then.  Though maybe that's an area for improvement in the future - we could use something more reliable than the tag down the road.

File: controllers/solrcloud_controller.go [hunk 5]
```diff
@@ -699,6 +695,41 @@ func reconcileCloudStatus(r *SolrCloudReconciler, solrCloud *solr.SolrCloud, log
 	return outOfDatePods, outOfDatePodsNotStarted, availableUpdatedPodCount, nil
 }
 
+func isPodReadyForBackup(pod corev1.Pod, backupOptions *solr.SolrBackupRestoreOptions) bool {
```
[the remark this answers is not in the document]
Comment (gerlowskija) [reply controllers/solrcloud_controller.go:698]: As written, this method checks if a single pod is ready to handle a backup for any/all of its repositories.  This information gets surfaced to consumers under the `backupRestoreReady` boolean field of `SolrCloudStatus`.

We can definitely change it to take a specific repository, but since this call isn't happening at backup-time, we'd still have to call it in a loop for each repository defined on the solrcloud instance and then "AND" each of those results together into the single `backupRestoreReady` value.  So I'm not really sure that gets us anything?

Unless you're thinking that we'd change SolrCloudStatus to include a distinct "backupReadiness" status for each repository so that we could check the one we care about at "backup time".  I'd be OK with that if that's what you're suggesting, just wanted to double check we're envisioning the same thing here.

**Comment by gerlowskija** [comment]
Thanks!

> Mind if I make some changes while you are working on other stuff?

Not at all, thanks for the help!  My plate cleared up a bit after posting the other day, so I'll actually be picking this up again in the short term (until paternity takes me away again).  So if I'm unable to finish, hopefully I can at least whittle down what'll be left when you turn your attention to it.

File: api/v1beta1/solrcloud_types.go [hunk 6]
```diff
@@ -325,20 +329,158 @@ type SolrEphemeralDataStorageOptions struct {
...
+	// Other options are to use a NFS volume.
+	Volume *corev1.VolumeSource `json:"volume,omitempty"`
+
+	// Select a custom directory name to mount the backup/restore data from the given volume.
+	// If not specified, then the name of the solrcloud will be used by default.
+	// +optional
+	Directory string `json:"directory,omitempty"`
+}
+
+func (managedRepository *ManagedStorage) IsManaged() bool { return true }
```
[the remark this answers is not in the document]
Comment (gerlowskija) [reply api/v1beta1/solrcloud_types.go]: AFAICT, golang "receiver functions" need to be defined in the same package as the type they're received by.  So I was able to move these out to a new file (solrcloud_type_helpers.go), but it's still in the API package by necessity.

Hope that's what you were looking for!

**Comment by gerlowskija** [comment]
Alright, quick update on my progress here.  Since I last posted, I've

- addressed all of Houston's requests (except two where I asked for clarification on what he wanted done)
- fixed a bug mentioned earlier around updating backup status when persistence was skipped.
- added some unit tests
- added examples (under `examples/`) and re-wrote the backup docs (in `docs/solr-backup/README.md`)
- done misc cleanup (e.g. removing personal debug logging, etc.)

That's everything I had planned to do here, unless @HoustonPutman has other feedback or gets a chance to see my responses from his earlier review.  But otherwise I think this is ready to go?

**Comment by gerlowskija** [comment]
Actually - I see there are a few errors in the PR build that come out of 'make lint'.  It looks like the 'lint' output proposes a fix in each line that has bad indentation etc. which is really cool.  Just wanted to check before I fix those: is there a programmatic way to apply those fixes, or is "by hand" the best approach for these 'make lint' errors?

**Comment by HoustonPutman** [comment]
@gerlowskija you can run ‘make fmt’ to fix the formatting issues or ‘make prepare’ to auto-fix all possible linting issues that can be auto-fixed

**Comment by gerlowskija** [comment]
Thanks Houston - I saw that 'make lint' advises using 'go fmt' to autocorrect issues as well.  Lots of options it looks like haha.

File: api/v1beta1/solrcloud_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply api/v1beta1/solrcloud_types.go]: Yes, exactly. We do that already for the SolrCloud and the SolrPrometheusExporter through `WithDefaults()` methods.

File: controllers/solrcloud_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply controllers/solrcloud_controller.go:697]: Yeah I was thinking that the SolrCloud status could start providing a list of ready backup Repos. If you don't think this is very feasible that's fine. We can always add it later.

File: controllers/util/solr_util.go [hunk 7]
```diff
@@ -215,16 +216,63 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, solrCloudStatus *solr.SolrCl
...
+			}
+		}
+
+		// Lastly, handle the volumes needed for any GCS credentials.
+		if backupOptions.GcsRepositories != nil {
+			for _, gcsRepository := range *backupOptions.GcsRepositories {
+				fals := false
+				solrVolumes = append(solrVolumes, corev1.Volume{
+					Name: gcsRepository.GetVolumeName(),
+					VolumeSource: corev1.VolumeSource{
```
Comment (HoustonPutman) [inline controllers/util/solr_util.go:260]: We might need to worry about restarting the Solr Pods when these secrets update. But again we can have a separate issue/PR for that.

**Comment by gerlowskija** [comment]
> Instead of lists for each individual repository type, we structure it more like Volumes

Houston and I discussed this a bit offline and I'm 👍 on the alternate syntax he proposed.

**Comment by HoustonPutman** [comment]
Ok I have refactored to use a new more kube-style way of listing repositories. I've also updated the documentation and examples.

One big difference is we are no longer supporting the legacy location for the backupRepository information. The SolrCloud.withDefaults() will auto-move this information to the `backupRepositories` list. However, we lose the old location where these directories were mounted. This is not an issue, since the Solr Operator will use this new location after it's upgrade. The only issue is when the Solr Operator has upgraded, but the SolrClouds are still in a rolling-restart to update the mountPath of the legacy backupRepo. Users will have to hold off of doing any backups until the SolrClouds they are using have been fully updated. Once this is done, there should be no issues with backwards-compatibility.

I have added a warning for this in the upgrade-notes for `v0.5.0`. I think it's an acceptable tradeoff to not need a lot of custom code. Also we will be able to remove the deprecated option completely in `v0.6.0` without creating an issue for existing clouds (since the deprecated option will have been moved to `backupRepositories`).

I think there are a few things left to do:

- [ ] Add helm chart options for the backupRepositories, update Helm documentation
- [ ] Add a changelog entry.
- [ ] Add assert messages in new tests.

Will try to wrap this up tomorrow. And hopefully do some integration tests afterwards. Would love help with the integration tests if anyone has time for that.

**Review by HoustonPutman — APPROVED** [review]
Ok I think this is good to go for now.

Will try to do some integration testing early next week, but there is definitely time to let this bake in and fix things before the next release.

Awesome work @gerlowskija , thanks for the great contribution! This makes the SolrBackup CRD far more usable, and will hopefully make the migration of SolrClouds to Kubernetes that much easier.

**Comment by gerlowskija** [comment]
Thanks a ton for seeing this through in my absence @HoustonPutman .  Will be awesome to see this in 0.5.0!
